[BEAM-3291] Add Kinesis write transform

aromanenko-dev opened the pull request:
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This PR introduces the support of write functionality to Kinesis through `KinesisIO.write()` call. 
The implementation is based on Amazon KPL which provides high throughput write access to Kinesis streams. Amazon KPL is licensed under the Amazon Software License so KPL jars can't  be distributed along with Apache Beam distribution. More details about Amazon KPL can be found here:
[https://docs.aws.amazon.com/streams/latest/dev/developing-producers-with-kpl.html](url)

I tried to keep new API as much as similar to already existed `KinesisIO.read()` API despite of the fact that it's based on AWS SDK for Kinesis. Also, I merged ITs of `KinesisIO.read()` and `KinesisIO.write()` into new one IT to guarantee that we can properly read the data that was written before by Beam KinesisIO.

aromanenko-dev commented:
Hi @jkff @pawel-kaczmarczyk @jbonofre @iemejia , can you please take a look?

jbonofre commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/BasicKinesisProvider.java:
```diff
@@ -0,0 +1,90 @@
+import javax.annotation.Nullable;
+
+/** Basic implementation of {@link AWSClientsProvider} used by default in {@link KinesisIO}. */
+class BasicKinesisProvider implements AWSClientsProvider {
```
Visibility is default, is it expected ?

aromanenko-dev replied:
Yes, it's supposed to be used only inside this package.

jbonofre commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +339,210 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+        config.setCredentialsRefreshDelay(100);
+
+        // Init Kinesis producer
+        producer = spec.getAWSClientsProvider().createKinesisProducer(config);
```
Is it OK to create the producer in `Setup` instead of `StartBundle` ? For instance, in streaming mode, the producer can be open for a very long time. I wonder if Kinesis peer can "close" a producer open for a long time.
Thoughts ?

aromanenko-dev replied:
Good point, @jbonofre. Actually, `KinesisProducer` (KP) is just a wrapper around native child process which is responsible for all communications with Kinesis endpoint. It's expected that only one instance of KP should be used per worker, creating new instance of KP is quite heavy operation (since it creates new system process) and all retries and error handling is incapsulated inside KP. I think, it would make sense to run long IT to see how it works under this condition.

jbonofre replied:
Agree as it's a heavy operation. I would like to check that we don't have any timeout or kick out from Kinesis with a long pipeline execution (like in streaming mode for instance). Thanks !

aromanenko-dev replied:
@jbonofre I performed a simple test to test this behaviour:
1) I created a pipeline that reads data from Kafka and writes into Kinesis 
2) I run this pipeline with direct runner and sent new message to Kafka.
3) I checked that new records was stored into Kinesis stream
4) Then I was waiting for 2 hours and send another message to Kafka.
5) After I checked that another record was stored in Kinesis too.

So, for now, I think we can conclude that it's not an issue. Are you agree?

jbonofre replied:
It sounds good to me. Thanks !

aromanenko-dev commented:
@echauchot, I'll answer and comment here since previous PR was closed.

> is there any embedded version of kinesis provided by Amazon for the tests?

Unfortunately, no. Only third-party implementation, for example [https://github.com/localstack/localstack](url) and [https://github.com/mhart/kinesalite](url)

> yes too early is what I meant, there is nothing related to backend resilience. I've seen cases (see https://issues.apache.org/jira/browse/BEAM-2541) where users create the graph on their dev machines that do not have access to the backend and then deploy and run pipeline in dataflow. expand() is invoked at construction time, you need to do the check at runtime. I had this discussion with @jkff as well on ES5 PR

I've got your point and agree with it. So yes, it makes sense to move this check into `@setup`.

> this KinesisRecordToTestData class is both in kinesis write test and in kinessis read test

Yes, I'll fix this. Good catch, thanx!

aromanenko-dev commented:
@echauchot I fixed two issue that you requested. Please, take a look on this https://github.com/apache/beam/pull/4468/commits/80549d754c85b93c84c09280b37ff3ed1689fdd9

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+      abstract Write build();
+    }
+
+    public Write withStreamName(String streamName) {
```
Please add javadocs for the builder methods, explaining nontrivial aspects (e.g. which properties are required, which have a default value and what it is, etc.)

aromanenko-dev replied:
Ok, sure

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+        Futures.addCallback(f, new UserRecordResultFutureCallback());
+      }
+
+      @FinishBundle
```
Does anything need to be done in `StartBundle`?

aromanenko-dev replied:
I don't think so since we create producer in `Setup`. Perhaps, in `StartBundle`, we can check  if it's not null (I'm not sure that it's possible) and create new instance if needed. What do you think?

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
-        return clientBuilder.build();
+
+        ListenableFuture<UserRecordResult> f =
+            producer.addUserRecord(spec.getStreamName(), partitionKey, explicitHashKey, data);
```
Are there any issues if a bundle has a very large number of elements?

aromanenko-dev replied:
Yes, potentially we can have OoM here because for every processed record producer adds new `SettableFuture<UserRecordResult>` into `ConcurrentHashMap` which actually contains outgoing records. So, probably, we can check number of outgoing messages before adding new user record and wait until it will be less than some maximum (configurable) number?

aromanenko-dev replied:
@jkff I added additional "on demand" flush which is called once the number of non-flushed records reaches maximum limit. It should prevent us from having large futures buffer.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        // Flush all outstanding records, blocking call
```
I'd recommend merging these two methods into one - I looked at `flushSync` and was very alarmed that it silently gives up when out of retries, only to find that the error reporting is done in the next method.

aromanenko-dev replied:
Yes, it makes sense because these two methods are called only from `finishBundle`.

aromanenko-dev replied:
Merged into one method.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+
+        @Override
+        public void onFailure(Throwable t) {
+          LOG.error("An exception occurred while processing a record", t);
```
Users don't look at error logs unless their pipeline is failing - errors must lead to exceptions, and a pipeline must not succeed unless it has successfully performed all the requested processing.

See https://beam.apache.org/contribute/ptransform-style-guide/#error-handling - specifically the part with "Never, ever, ever do this" :)

Please make sure that flushing throws these errors. See implementation of BigtableIO.write() for a similar pattern.

echauchot replied:
+1 I left similar comment

aromanenko-dev replied:
Yes, I did what I should never have to do =) Thank you for pointing out this. Actually, I think we need to do the same (throw an exception) in `onSuccess()` too in case if `result.isSuccessful() == false`. Are you agree with this?

echauchot replied:
yes, throw an unchecked exception

aromanenko-dev replied:
@echauchot @jkff I added a logic to throw an exception in case of failed records (it's based on BigtableIO code).

echauchot commented:
from my comments en the previous PR:
- on asynchronous batch write:

> producer.addUserRecord is asynchronous call which adds record the queue which then will be grouped into batches and send to backend in background (Kinesis producer forks native processes to do this job). So, this is a responsibility of KPL in this case. Client can change number of records per batch with producer property (see withProducerProperties()).
> 
> What kind of cache mechanism do you mean?
The behavior you describe looks good to me

- on the assertEquals for with...

> IMHO it makes sense to test that with... works as expected.
I is equivalent to testing autovalue which is already tested. No other IO test that

aromanenko-dev commented:
@jkff Thank you for review, I added some comments and questions.

aromanenko-dev commented:
@echauchot 
> on the assertEquals for with...
>> IMHO it makes sense to test that with... works as expected.
>>> I is equivalent to testing autovalue which is already tested. No other IO test that

Actually, `BigtableIOTest` does this (which I used as inspiration for my tests) in `testReadBuildsCorrectly` and `testWriteBuildsCorrectly`. Since, `with...` methods can be not only just setters then I'd propose to leave this test.

echauchot commented:
> Actually, BigtableIOTest does this (which I used as inspiration for my tests) in testReadBuildsCorrectly and testWriteBuildsCorrectly. Since, with... methods can be not only just setters then I'd propose to leave this test.
we can keep it. Fair enough

aromanenko-dev commented:
@jkff I pushed two commits that address your review comments. Please, take a look on this.

iemejia commented:
retest this please

iemejia commented:
Run Java Gradle PreCommit

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+
+    /**
+     * Specify the number of retries that will be used to flush the outstanding records in
+     * case if they were not flushed from the first time. Default number of retries is
```
I'd recommend to not allow a user to configure this, for all the usual reasons why other transforms in Beam prefer not to expose tuning knobs. The default value is likely good enough for all cases.

aromanenko-dev replied:
Well, I agree that if it's a default rule for Beam transforms then it makes sense to drop it. But from the user point of view, I'd prefer to have some knobs in case when something goes not as expected.

jkff replied:
I understand that in general this is up to debate, but in Beam the rule of thumb is to expose a tuning knob only as a last resort: if someone could conceivably end up in a situation where their pipeline exhibits catastrophically poor behavior with the default value AND if it seems infeasible to choose a proper value automatically in general. I don't think this applies here - so for consistency with the rest of Beam codebase, I'd like to ask to remove this knob.

For context, the justification for this rule of thumb is that we've found that, whenever there is a tuning knob, nearly all users set it to a bad value - or perhaps to a value that used to be good at the moment they chose it, but not anymore (e.g. because the pipeline changed, the data changed, implementation of underlying libraries changed, the runner changed, etc) - and the user doesn't have a process in place to keep regularly manually re-tuning the parameter for every use case (I've never seen a user who has such a process).

E.g. inside Google, the legacy MapReduce used to have hundreds of tuning parameters, and when a user complained about a performance problem, often the first recommendation was "have you tried removing all tuning parameters" and it very often helped a great deal :) When a value is explicitly specified, we have to respect it and can't automatically make things better.

aromanenko-dev replied:
Thank you for so detailed explanation, it makes sense for me.  I agree to not provide this knob to end user but I'd keep it for using in unit test - I want to test throwing an exception when flush is failing and I want to reduce a number of retries to minimum, otherwise this UT takes long time to finish. 
So, I changed the visibility of `withRetries()` to package instead of public (like `withBigtableService()` in `BigtableIO`). 
Are you ok with this?

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+        int retries = spec.getRetries();
+        int numOutstandingRecords = producer.getOutstandingRecordsCount();
+
+        while (numOutstandingRecords > numMax && retries-- > 0) {
```
In what case would `producer.flush()` not flush everything - why do we even need a retry loop? I think new records cannot be added while this runs because they're only added from `ProcessElement`.

aromanenko-dev replied:
Let me explain how actually `KinesisProducer` flushes records. When we create new instance of `KinesisProducer`, it spawns new child process (this is a native process) which, in fact, does all work to aggregate and to send the records to Kinesis. In its order, `KinesisProducer` "talks" to this process through pipes. 

So, once we call `producer.flush()` then it just sends a command to child process to flush outstanding records but with no guarantee that the buffered records will be sent. For this, we need to run a loop to check the number of outstanding records to make sure that all or required number records were sent.

Actually, `KinesisProducer` has another method - `producer.flushSync()` - which just runs infinite loop until all messages will be flushed:
```java
    public void flushSync() {
        while (getOutstandingRecordsCount() > 0) {
            flush();
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) { }
        }
    }
```
but I'm not sure it's a good idea to use this method to flush everything in `TearDown` or `FinishBundle` since it might block forever. That is why I created a new method which tries to flush only the number of retry times and it will throw an exception in case if it still has **more than required** number of outstanding records and I use it to flush all bundle records in `FinishBundle` (`flushAll()` equal to `flush(0)`).

Also, I call this method `flush(MAX_NUM_RECORDS)` in `ProcessElement` to prevent unlimited growing the buffer of added futures that are stored in `HashMap` (as we discussed it before). The goal of this call is to wait until number of outgoing messages becomes less than `MAX_NUM_RECORDS` and yes, until this, we can't add new records. In the first version of PR, I didn't have this call but, potentially, we could have OOM there, so it was added.

jkff replied:
Oh I see, I didn't realize flush was asynchronous. In that case the current code is fine, modulo removal of the tuning knob (I think it should be sufficient to use a default of e.g. retrying for up to 1 minute - I can't think of a case where it would be critically important to use a different value; I'm happy to reconsider if such a case emerges in practice).

aromanenko-dev replied:
Ok, I changed a total time of retrying to 1 minute. I believe it should quite enough for most cases as a default value.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+                  numOutstandingRecords,
+                  numMax);
+          LOG.error(message);
+          throw new IOException(message);
```
I'm not sure we should fail here. We might be able to proceed without losing data; if an OOM happens, there'll be plenty of logging to explain why.

aromanenko-dev replied:
As I mentioned before, I use this method also to flush all bundle records in `FinishBundle`, so it makes sense to fail. Probably, we can change this logic and fail only in `FinishBundle` (or even in `TearDown`). 
What do you think?

jkff replied:
Oh right. It would make sense to fail only in `FinishBundle`, but probably not worth the trouble.

aromanenko-dev replied:
Yes, let's keep it "as it is" for the moment.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+
+        String message =
+            String.format(
+                "Some errors occurred writing to Kinesis. First %d errors: %s",
```
Seems that we actually print all errors, not the first N.

aromanenko-dev replied:
I use `LinkedBlockingDeque` with fixed capacity (N) to store failure exception. So, it keeps only first N failures, others will be ignored once we add then using `offer()` method.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+                "Some errors occurred writing to Kinesis. First %d errors: %s",
+                i,
+                logEntry.toString());
+        LOG.error(message);
```
It's not necessary to both log and throw an error - an exception will surely be logged at a higher level.

aromanenko-dev replied:
Ok, agree.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+
+        @Override public void onSuccess(UserRecordResult result) {
+          if (!result.isSuccessful()) {
+            failures.add(new KinesisWriteException("Put record was not successful."));
```
* Why is the one above `offer()` but this one `add()`?
* Is it possible to log what the error was?
* Is it possible (and is it useful?) to log something identifying which particular record failed?  Also applies to `onFailure`.

aromanenko-dev replied:
Thank you for catching this, I'll fix it and add some additional error details.

aromanenko-dev commented:
@jkff Thank you for another review iteration. I addressed some of your comments and updated my PR accordingly. Also, I replied to your other comments, please, take a look.

iemejia commented:
After the retry fixes plus hiding the numRetries from the public API it seems to me almost finished. What do you think @jkff, ready to merge ?

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +337,320 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+     * case if they were not flushed from the first time. Default number of retries is
+     * {@code DEFAULT_NUM_RETRIES = 10}.
+     *
+     * <p>This is used for testing.
```
Please mark it `@VisibleForTesting`.

aromanenko-dev replied:
Sure, thanks!

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +337,320 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+      public void finishBundle() throws Exception {
+        checkForFailures();
+
+        // Flush all outstanding records, blocking call
```
I think this should be swapped with checkForFailures - new errors might appear during flushing.

jkff replied:
OK actually this comment is not minor. Please fix :)

aromanenko-dev replied:
Yes, I agree it makes sense to change an order.

jkff commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -272,52 +337,320 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+          // wait until outstanding records will be flushed
+          Thread.sleep(retryTimeout);
+          numOutstandingRecords = producer.getOutstandingRecordsCount();
+          retryTimeout *= 2; // exponential backoff
```
Seems that with the default number of 10 retries, we'll wait for up to 2047 seconds (1 + 2 + .. + 1024) ? Fine with me, just wanted to check that it's intentional.

aromanenko-dev replied:
Actually, in my last commit (a43afb9f3a09086f20ace8471b9717a031f733f1), I reduced default number of retries to 6, so the total time of retry is about 1 minute. I believe it should be fine.
